File: app/src/main/java/net/vornao/ddns/dome/handler/DeviceHandler.java
package net.vornao.ddns.dome.handler;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;

import com.google.gson.Gson;
import com.google.gson.JsonObject;

import net.vornao.ddns.dome.callbacks.UpdateDeviceCallback;
import net.vornao.ddns.dome.shared.Const;
import net.vornao.ddns.dome.devices.Device;
import net.vornao.ddns.dome.devices.EnergyMonitor;
import net.vornao.ddns.dome.devices.RemoteSwitch;
import net.vornao.ddns.dome.devices.Thermostat;

import org.eclipse.paho.android.service.MqttAndroidClient;
import org.eclipse.paho.client.mqttv3.IMqttActionListener;
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.IMqttToken;
import org.eclipse.paho.client.mqttv3.MqttCallback;
import org.eclipse.paho.client.mqttv3.MqttClient;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;

import java.util.Arrays;
import java.util.HashMap;

public class DeviceHandler {

    @NonNull
    private final MqttAndroidClient client;
    @NonNull
    private final MqttConnectOptions mqttConnectOptions;
    @NonNull
    private final MqttCallback callback;
    @NonNull
    private final DeviceHandler deviceHandler;

    private String house_id;


    private String houseTopic;
    private final UpdateDeviceCallback updateDeviceCallback;

    private final Gson gson = new Gson();

    @NonNull
    public static HashMap<Integer, Device> devices = new HashMap<>();



    public DeviceHandler(Context context, String server, String username, @NonNull String pass,
                         @NonNull HashMap<Integer, Device> devices, String house_id, UpdateDeviceCallback udc) {

        client = new MqttAndroidClient(context, server, MqttClient.generateClientId());
        updateDeviceCallback = udc;

        this.house_id = house_id;
        DeviceHandler.devices = devices;

        houseTopic = "dome/" + house_id;
        Log.i("MQTT", houseTopic);
        this.deviceHandler = this;

        mqttConnectOptions = new MqttConnectOptions();
        mqttConnectOptions.setUserName(username);
        mqttConnectOptions.setPassword(pass.toCharArray());

        callback = new MqttCallback() {

            @Override
            public void connectionLost(Throwable cause) {
                Log.i("MQTT", "CONNLOST");
            }

            @Override
            public void messageArrived(@NonNull String topic, @NonNull MqttMessage message) {

                JsonObject recv = gson.fromJson(new String(message.getPayload()), JsonObject.class);
                String type =  recv.get(Const.DEVICE_TYPE).getAsString();
                int tempId  =  recv.get(Const.DEVICE_ID).getAsInt();

                if (topic.equals(houseTopic)) {

                    if (!devices.containsKey(tempId)) {
                        switch (type) {
                            case Const.THERMOSTAT: addThermostat(tempId, recv); break;
                            case Const.REMOTE_SWITCH: addRemoteSwitch(tempId, recv); break;
                            case Const.ENERGY_MONITOR: addEnergyMonitor(tempId, recv); break;
                        }
                    }else {
                        switch (type) {
                            case Const.THERMOSTAT: updateThermostat(tempId, recv); break;
                            case Const.REMOTE_SWITCH: updateRemoteSwitch(tempId, recv); break;
                            case Const.ENERGY_MONITOR: updateEnergyMonitor(tempId, recv); break;
                        }
                    }
                    updateDeviceCallback.onSuccess(tempId);
                }
            }

            @Override
            public void deliveryComplete(IMqttDeliveryToken token) {
            }
        };
    }

    public void connect() {
        try {
            IMqttToken token = client.connect(mqttConnectOptions);
            token.setActionCallback(new IMqttActionListener() {

                @Override
                public void onSuccess(IMqttToken asyncActionToken) {
                    Log.i("MQTT", "SUCCESS");
                    Log.d("MQTT", houseTopic);
                    subscribe(houseTopic);
                    client.setCallback(callback);
                    updateDeviceCallback.onConnSuccess();
                }

                @Override
                public void onFailure(IMqttToken asyncActionToken, Throwable exception) {
                    Log.i("MQTT", "FAIL");
                    updateDeviceCallback.onFailure();
                }
            });

        } catch (MqttException e) {
            Log.i("MQTT", "ERROR CONNECTING TO MQTT BROKER");
            e.printStackTrace();
        }
    }

    public void disconnect() {
        try {
            IMqttToken token = client.disconnect();
            token.setActionCallback(new IMqttActionListener() {
                @Override
                public void onSuccess(IMqttToken asyncActionToken) {}
                @Override
                public void onFailure(IMqttToken asyncActionToken, Throwable exception) {
                }
            });

        } catch (MqttException e) {
            e.printStackTrace();
        }
    }

    public void close() {
        try {
            //client.disconnect();
            client.close();
        } catch (Exception e) {
            Log.d("MQTT", Arrays.toString(e.getStackTrace()));
        }
    }
       

    public void publish(String topic, @NonNull String payload, int qos){
        try {

            client.publish(topic, new MqttMessage(payload.getBytes()));

        }catch (@NonNull MqttException | NullPointerException e){
            Log.i("MQTT", e.getLocalizedMessage());
        }
    }

    public void subscribe(String topic){
        try{
            client.subscribe(topic, 0);
        }
        catch (@NonNull MqttException | NullPointerException e){
            Log.i("MQTT", "FAILED TO SUBSCRIBE");
        }
    }

    private void addThermostat(int id, @NonNull JsonObject recv){
        devices.put(id, new Thermostat(id, Const.THERMOSTAT, deviceHandler, house_id, "Thermostat"));
        updateDeviceCallback.onNewDevice(id);
        updateThermostat(id, recv);
    }

    private void addRemoteSwitch(int id, @NonNull JsonObject recv){
        devices.put(id, new RemoteSwitch(id, Const.REMOTE_SWITCH, deviceHandler, house_id, "Remote switch"));
        updateRemoteSwitch(id, recv);
        updateDeviceCallback.onNewDevice(id);
    }

    private void addEnergyMonitor(int id, @NonNull JsonObject recv){
        devices.put(id, new EnergyMonitor(id, Const.ENERGY_MONITOR, deviceHandler, house_id, "Power meter"));
        updateDeviceCallback.onNewDevice((id));
        updateEnergyMonitor(id, recv);
    }

    private void updateThermostat(int id, @NonNull JsonObject recv){
        ((Thermostat) devices.get(id)).setHum(recv.get(Const.THERM_HUM_KEY).getAsFloat());
        ((Thermostat) devices.get(id)).setTemp(recv.get(Const.THERM_TEMP_KEY).getAsFloat());
        ((Thermostat) devices.get(id)).setThreshold(recv.get(Const.THERM_THRESHOLD_KEY).getAsFloat());
        ((Thermostat) devices.get(id)).setMode(recv.get(Const.STATUS_KEY).getAsInt());
        devices.get(id).setOnline();

    }

    private void updateRemoteSwitch(int id, @NonNull JsonObject recv){
        ((RemoteSwitch) devices.get(id)).setMode(recv.get(Const.STATUS_KEY).getAsInt());
        devices.get(id).setOnline();
    }

    private void updateEnergyMonitor(int id, @NonNull JsonObject recv){
        ((EnergyMonitor)devices.get(id)).setDataUrl(recv.get(Const.EMON_DATA_PATH_KEY).getAsString());
        ((EnergyMonitor)devices.get(id)).setPower(recv.get(Const.EMON_POWER).getAsFloat());
        devices.get(id).setOnline();
    }

    public String getHouseTopic() {
        return houseTopic;
    }

    public void setHouseTopic(String houseTopic) {
        this.house_id = houseTopic;
        this.houseTopic = "dome/" + house_id;
    }
}

File: app/src/main/java/net/vornao/ddns/dome/handler/DeviceHandler.kt
package net.vornao.ddns.dome.handler

import android.content.Context
import android.util.Log

import net.vornao.ddns.dome.callbacks.UpdateDeviceCallback
import org.eclipse.paho.android.service.MqttAndroidClient
import org.eclipse.paho.client.mqttv3.MqttConnectOptions
import org.eclipse.paho.client.mqttv3.MqttCallback
import com.google.gson.Gson
import org.eclipse.paho.client.mqttv3.IMqttToken
import org.eclipse.paho.client.mqttv3.IMqttActionListener
import org.eclipse.paho.client.mqttv3.MqttException
import org.eclipse.paho.client.mqttv3.MqttMessage
import com.google.gson.JsonObject
import net.vornao.ddns.dome.devices.Device
import net.vornao.ddns.dome.devices.Thermostat
import net.vornao.ddns.dome.devices.RemoteSwitch
import net.vornao.ddns.dome.devices.EnergyMonitor
import net.vornao.ddns.dome.shared.Const
import org.eclipse.paho.client.mqttv3.MqttClient
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken
import java.lang.Exception
import java.lang.NullPointerException
import java.util.*

class DeviceHandler(context: Context?, server: String?, username: String?, pass: String,
                    devices: HashMap<Int, Device>, private var house_id: String, udc: UpdateDeviceCallback) {
    private val client: MqttAndroidClient = MqttAndroidClient(context, server, MqttClient.generateClientId())
    private val mqttConnectOptions: MqttConnectOptions
    private val callback: MqttCallback
    private val deviceHandler: DeviceHandler
    var houseTopic: String

    private val updateDeviceCallback: UpdateDeviceCallback = udc
    private val gson = Gson()
    fun connect() {
        try {
            val token = client.connect(mqttConnectOptions)
            token.actionCallback = object : IMqttActionListener {
                override fun onSuccess(asyncActionToken: IMqttToken) {
                    Log.i("MQTT", "SUCCESS")
                    Log.d("MQTT", houseTopic)
                    subscribe(houseTopic)
                    client.setCallback(callback)
                    updateDeviceCallback.onConnSuccess()
                }

                override fun onFailure(asyncActionToken: IMqttToken, exception: Throwable) {
                    Log.i("MQTT", "FAIL")
                    updateDeviceCallback.onFailure()
                }
            }
        } catch (e: MqttException) {
            Log.i("MQTT", "ERROR CONNECTING TO MQTT BROKER")
            e.printStackTrace()
        }
    }

    fun close() {
        try {
            //client.disconnect();
            client.close()
        } catch (e: Exception) {
            Log.d("MQTT", Arrays.toString(e.stackTrace))
        }
    }

    fun publish(topic: String?, payload: String, qos: Int) {
        try {
            client.publish(topic, MqttMessage(payload.toByteArray()))
        } catch (e: MqttException) {
            Log.i("MQTT", e.localizedMessage)
        } catch (e: NullPointerException) {
            Log.i("MQTT", e.localizedMessage)
        }
    }

    fun subscribe(topic: String?) {
        try {
            client.subscribe(topic, 0)
        } catch (e: MqttException) {
            Log.i("MQTT", "FAILED TO SUBSCRIBE")
        } catch (e: NullPointerException) {
            Log.i("MQTT", "FAILED TO SUBSCRIBE")
        }
    }

    private fun addThermostat(id: Int, recv: JsonObject) {
        devices[id] = Thermostat(id, Const.THERMOSTAT, deviceHandler, house_id, "Thermostat")
        updateDeviceCallback.onNewDevice(id)
        updateThermostat(id, recv)
    }

    private fun addRemoteSwitch(id: Int, recv: JsonObject) {
        devices[id] = RemoteSwitch(id, Const.REMOTE_SWITCH, deviceHandler, house_id, "Remote switch")
        updateRemoteSwitch(id, recv)
        updateDeviceCallback.onNewDevice(id)
    }

    private fun addEnergyMonitor(id: Int, recv: JsonObject) {
        devices[id] = EnergyMonitor(id, Const.ENERGY_MONITOR, deviceHandler, house_id, "Power meter")
        updateDeviceCallback.onNewDevice(id)
        updateEnergyMonitor(id, recv)
    }

    private fun updateThermostat(id: Int, recv: JsonObject) {
        (devices[id] as Thermostat?)!!.hum = recv[Const.THERM_HUM_KEY].asFloat
        (devices[id] as Thermostat?)!!.temp = recv[Const.THERM_TEMP_KEY].asFloat
        (devices[id] as Thermostat?)!!.threshold = recv[Const.THERM_THRESHOLD_KEY].asFloat
        (devices[id] as Thermostat?)!!.mode = recv[Const.STATUS_KEY].asInt
        devices[id]!!.setOnline()
    }

    private fun updateRemoteSwitch(id: Int, recv: JsonObject) {
        (devices[id] as RemoteSwitch?)!!.mode = recv[Const.STATUS_KEY].asInt
        devices[id]!!.setOnline()
    }

    private fun updateEnergyMonitor(id: Int, recv: JsonObject) {
        (devices[id] as EnergyMonitor?)!!.dataUrl = recv[Const.EMON_DATA_PATH_KEY].asString
        (devices[id] as EnergyMonitor?)!!.power = recv[Const.EMON_POWER].asFloat
        devices[id]!!.setOnline()
    }


    companion object {
        var devices = HashMap<Int, Device>()
    }

    init {
        Companion.devices = devices
        houseTopic = "dome/$house_id"
        Log.i("MQTT", houseTopic)
        deviceHandler = this
        mqttConnectOptions = MqttConnectOptions()
        mqttConnectOptions.userName = username
        mqttConnectOptions.password = pass.toCharArray()
        callback = object : MqttCallback {
            override fun connectionLost(cause: Throwable) {
                Log.i("MQTT", "CONNLOST")
            }

            override fun messageArrived(topic: String, message: MqttMessage) {
                val recv = gson.fromJson(String(message.payload), JsonObject::class.java)
                val type = recv[Const.DEVICE_TYPE].asString
                val tempId = recv[Const.DEVICE_ID].asInt
                Log.d("MQTT", houseTopic)
                if (topic == houseTopic) {
                    if (!devices.containsKey(tempId)) {
                        when (type) {
                            Const.THERMOSTAT -> addThermostat(tempId, recv)
                            Const.REMOTE_SWITCH -> addRemoteSwitch(tempId, recv)
                            Const.ENERGY_MONITOR -> addEnergyMonitor(tempId, recv)
                        }
                    } else {
                        when (type) {
                            Const.THERMOSTAT -> updateThermostat(tempId, recv)
                            Const.REMOTE_SWITCH -> updateRemoteSwitch(tempId, recv)
                            Const.ENERGY_MONITOR -> updateEnergyMonitor(tempId, recv)
                        }
                    }
                    updateDeviceCallback.onSuccess(tempId)
                }
            }

            override fun deliveryComplete(token: IMqttDeliveryToken) {}
        }
    }
}